pandas/core/generic.py

```diff
@@ -8832,6 +8837,16 @@ def _where(
         fill_value = bool(inplace)
         cond = cond.fillna(fill_value)
 
+        # With ArrayManager, `fillna` does not automatically change object dtype
+        # back to bools (if the alignment made it object by introducing NaNs).
+        # So in this case we cast back to bool manually *if* the original columns
+        # before aligning were bool
+        # TODO this workaround can be removed once we have nullable boolean dtype
+        # as default
+        if isinstance(cond, ABCDataFrame) and cond._has_array_manager:
+            if not all(is_bool_dtype(dt) for dt in cond.dtypes) and all_bool_columns:
+                cond = cond.astype(bool)
+
```

Comment:
is there a viable way to put these patches in with the ArrayManager code?

Reply:
I don't directly see one, as we don't want to change (IMO) the `fillna` behaviour. 
Unless we we would add a keyword for this (only for internal use) to the manager `fillna` method, and call `cond._mgr.fillna` here instead of `cond.fillna` to be able to pass it through, but not sure that would be any cleaner (as long as it is only used once, at least, there might come up other places where this is needed later)

Reply:
left same comment above. I think we are going to want to push *more* code to the manager (kind of the opposite that we have been doing of late though).